src/include/RADOS/RADOS.hpp

```diff
@@ -0,0 +1,933 @@
+// have to pay for the string copy to construct an object_t.
+
+class Object {
+  friend RADOS;
+  friend std::hash<Object>;
+
+public:
+  Object(std::string_view s);
+  Object(std::string&& s);
+  Object(const std::string& s);
+  ~Object();
+
+  Object(const Object& o);
+  Object& operator =(const Object& o);
+
+  Object(Object&& o);
+  Object& operator =(Object&& o);
+
+  operator std::string_view() const;
+
+  friend bool operator <(const Object& lhs, const Object& rhs);
+  friend bool operator <=(const Object& lhs, const Object& rhs);
+  friend bool operator >=(const Object& lhs, const Object& rhs);
+  friend bool operator >(const Object& lhs, const Object& rhs);
+
+  friend bool operator ==(const Object& lhs, const Object& rhs);
+  friend bool operator !=(const Object& lhs, const Object& rhs);
+
+private:
+
+  static constexpr std::size_t impl_size = 4 * 8;
+  std::aligned_storage_t<impl_size> impl;
+};
+
+// Not the same as the librados::IoCtx, but it does gather together
+// some of the same metadata. Since we're likely to do multiple
+// operations in the same pool or namespace, it doesn't make sense to
+// redo a bunch of lookups and string copies.
+
+struct IOContext {
+  friend RADOS;
+
+  IOContext();
+  explicit IOContext(std::int64_t pool);
+  IOContext(std::int64_t _pool, std::string_view _ns);
+  IOContext(std::int64_t _pool, std::string&& _ns);
+  ~IOContext();
+
+  IOContext(const IOContext& rhs);
+  IOContext& operator =(const IOContext& rhs);
+
+  IOContext(IOContext&& rhs);
+  IOContext& operator =(IOContext&& rhs);
+
+  std::int64_t pool() const;
+  void pool(std::int64_t _pool);
+
+  std::string_view ns() const;
+  void ns(std::string_view _ns);
+  void ns(std::string&& _ns);
+
+  // Because /some fool/ decided to disallow optional references,
+  // you'd have to construct a string in an optional which I would then
+  // take an optional reference to. Thus a separate 'clear' method.
+  std::optional<std::string_view> key() const;
+  void key(std::string_view _key);
+  void key(std::string&& _key);
+  void clear_key();
+
+  std::optional<std::int64_t> hash() const;
+  void hash(std::int64_t _hash);
+  void clear_hash();
+
+  std::optional<std::uint64_t> read_snap() const;
+  void read_snap(std::optional<std::uint64_t> _snapid);
+
+  // I can't actually move-construct here since snapid_t is its own
+  // separate class type, not an alias.
+  std::optional<
+    std::pair<std::uint64_t,
+	      std::vector<std::uint64_t>>> write_snap_context() const;
+  void write_snap_context(std::optional<
+			  std::pair<std::uint64_t,
+			              std::vector<std::uint64_t>>> snapc);
+private:
+
+  static constexpr std::size_t impl_size = 16 * 8;
+  std::aligned_storage_t<impl_size> impl;
+};
+
+inline constexpr std::string_view all_nspaces("\001"sv);
+
+enum class cmpxattr_op : std::uint8_t {
+  eq  = 1,
+  ne  = 2,
+  gt  = 3,
+  gte = 4,
+  lt  = 5,
+  lte = 6
+};
+
+enum class alloc_hint : uint32_t {
```

Comment:
`enum class` doesn't seem like a good fit for flags that are meant to be combined; calling set_alloc_hint() like this is a static_castastrophe:
```c++
auto hint = static_cast<alloc_hint>(static_cast<uint32_t>(alloc_hint::random_write) |
                                    static_cast<uint32_t>(alloc_hint::incompressible));
op.set_alloc_hint(len, len, hint);
```
i also worry about duplicating these defintions; we're likely to miss future additions to librados.h or vice versa

Reply:
I'll make it a pile of constants.

I'd rather duplicate it than not. Otherwise we end up a second class citizen having to dip into the legacy librados for everything.

Reply:
you could potentially pull them out of librados.h into a common header. or at least add a comment in librados.h and RADOS.hpp to note that changes need to be applied to both?

Reply:
otherwise, we need to define all the operators for it. like https://en.cppreference.com/w/cpp/types/byte